### DISCUSSION TITLE
How to shrink installation size?

### DISCUSSION THREAD

**Discussion by reitzig** [pyenv/pyenv#2868] [Q&A]
I noticed that Python installations installed with pyenv are _huge_ compared to the smallest functional setup I know.

Compare:

```shell
❯ du -sh Software/pyenv/versions/3.11.7/lib/python3.11
364M	Software/pyenv/versions/3.11.7/lib/python3.11
```

With:

```shell
❯ docker run --rm -it python:3.11-alpine du -sh /usr/local/lib/python3.11 
35.1M	/usr/local/lib/python3.11
```

That's not a trivial difference, especially since I'm up to building container images with multiple Python versions using pyenv. In fact, just the test folder of a pyenv installation (which doesn't even exist in the official Python image) is larger than the whole image:

```shell
❯ du -sh Software/pyenv/versions/3.11.7/lib/python3.11/test
123M	Software/pyenv/versions/3.11.7/lib/python3.11/test
❯ docker run --rm -it python:3.11-alpine du -sh /
58.2M	/
```

So I wonder: is there a way to strip down pyenv-based installations? How?

My first (ignorant) attempt was:

```shell
rm -rf "${PYENV_ROOT}"/versions/*/lib/python*/test
```

which immediately breaks `pip install`:

```
Traceback (most recent call last):
  File "/pyenv/versions/3.11.7/bin/pip", line 5, in <module>
    from pip._internal.cli.main import main
ModuleNotFoundError: No module named 'pip'
```

**Comment by native-api** [comment]
Pyenv installs a fully functional Python with (mostly) default options.
AFAICS, the default installation did grow in size significantly between 3.10 and 3.11 (for comparison, some other flavors):
```
$ du -sh ~/.pyenv/versions/*
207M	/home/vmuser/.pyenv/versions/3.10.13
305M	/home/vmuser/.pyenv/versions/3.11.6
307M	/home/vmuser/.pyenv/versions/3.12.0
895M	/home/vmuser/.pyenv/versions/graalpy-23.1.0
902M	/home/vmuser/.pyenv/versions/graalpy-community-23.1.0
653M	/home/vmuser/.pyenv/versions/miniconda3-3.11-23.9.0-0
```

As https://stackoverflow.com/questions/5138801/how-to-compile-python-from-source-and-get-a-clean-minimal-install suggests, you can check distro scripts on what they delete -- in this case, https://github.com/docker-library/python/blob/789d789e4a8db71d3d393667971c49b845ffdc3f/3.11/alpine3.19/Dockerfile . Note that they delete `.pyc` and `.pyo` -- which will be regenerated once you start executing stuff.

**Reply by reitzig** [reply]
Heh, that _is_ a rather obvious place to look, isn't it. Hindsight ... 😅 Thanks!

Copying [docker-library/python@789d789e:3.11/alpine3.19/Dockerfile#106](https://github.com/docker-library/python/blob/789d789e4a8db71d3d393667971c49b845ffdc3f/3.11/alpine3.19/Dockerfile#L106-L111) does indeed more than halve the size of each installation:

```shell
❯ du -sh /pyenv/versions/*/lib/python3*
94M	/pyenv/versions/3.11.7/lib/python3.11
105M	/pyenv/versions/3.12.1/lib/python3.12
```

 Still larger by a factor 2-3 than what they ship in their (`alpine` and `slim-bookworm`) images, but good enough for me at this point. 👌

**Reply by reitzig** [reply]
So, my next question _here_ is: Would it make sense to add an option like `pyenv install --slim` that does something similar to create a smaller installation than it does by default?

**Comment by native-api** [comment]
> So, my next question _here_ is: Would it make sense to add an option like `pyenv install --slim` that does something similar to create a smaller installation than it does by default?

As per the comment above, deleting `.pyc` and `.pyo` does not seem to be good UX: they will eventually be regenerated, reclaiming the "freed" space, and will only result in a slowdown while the interpreter is doing that.

I guess you can create a feature request and if it shows sufficient demand, we'll implement it as we do with other highly-requested features.

You can also implement this in a plugin, as a postinstall hook.
